Forbid destination files overwrite by default

karitra opened the pull request:
- forbid recovery overwrite of destination paths.
 - exceptions of BlobRepairer.fix command are stored in Json artifact.
 - minor refactorings.

karitra committed:
v0.1.2

shaitan commented on eblob_kit.py:
```diff
@@ -656,6 +657,7 @@ def copy_record(src, dst, header):
     """Copy record from @src to @dst specified by @header."""
     new_header = header.key + struct.pack('<4Q', header.flags.flags, header.data_size,
                                           header.disk_size, dst.data.file.tell())
+
```
```suggestion
```

shaitan commented on eblob_kit.py:
```diff
@@ -1041,10 +1043,14 @@ def check(self, verify_csum, fast):
         index_path = os.path.join(destination, basename + '.index')
+
+        if not overwrite_distance and os.path.exists(index_path):
+            raise RuntimeError("can't recover index to existant file {}".format(index_path))
```
```suggestion
            raise RuntimeError("can't recover index to existing file {}".format(index_path))
```

shaitan commented on eblob_kit.py:
```diff
@@ -1064,11 +1070,15 @@ def recover_index(data, destination):
-        blob = Blob.create(blob_path)
+
+        if not overwrite_distance and os.path.exists(blob_path):
+            raise RuntimeError("can't recover to already existing blob file: {}".format(blob_path))
```
Please, be consistent with similar messages.

shaitan replied:
It is still actual.

shaitan commented on eblob_kit.py:
```diff
@@ -1077,10 +1087,9 @@ def recover_blob(self, destination):
-                logging.error('I have faced with broken record which I can not skip.')
             if not header:
                 skipped_records += 1
+                logging.error('I have faced with broken record which I have to skip.')
```
Why is it done in this commit?

shaitan commented on eblob_kit.py:
```diff
@@ -1501,8 +1531,10 @@ def fix_blob_command(ctx, path, destination, noprompt):
 @click.option('-y', '--yes', 'noprompt', is_flag=True, default=False,
               help="Assume Yes to all queries and do not prompt, "
               "will be switched on when verbosity option not set")
+@click.option('-f', '--force', is_flag=True, default=False,
```
`force` -> `overwrite`

shaitan commented on eblob_kit.py:
```diff
@@ -407,10 +407,10 @@ def __init__(self, path, mode='rb'):
 
     @staticmethod
-    def create(path):
+    def create(path, mode='ab'):
```
Please, move this into separate commit

karitra committed:
refact: join two `if` branches with same condition

shaitan commented on eblob_kit.py:
```diff
@@ -61,6 +61,17 @@ def dump_to_file(file_name, results):
 
 
+def is_destination_writable(src_path, dst_path, overwrite=False):
+    """Check 'dst_path' file overwritable.
```
```suggestion
    """Check 'dst_path' file overwritability.
```

shaitan commented on eblob_kit.py:
```diff
@@ -1041,10 +1053,14 @@ def check(self, verify_csum, fast):
         basename = os.path.basename(data.path)
         index_path = os.path.join(destination, basename + '.index')
+
+        if not overwrite and os.path.exists(index_path):
```
Why isn't `is_destination_writable` used here?

shaitan commented on eblob_kit.py:
```diff
@@ -1064,11 +1079,15 @@ def recover_index(data, destination):
+        if not is_destination_writable(self._blob.data.path,  blob_path, overwrite):
+            raise RuntimeError("can't recover to already existing blob file: {}".format(blob_path))
+
+        blob = Blob.create(path=blob_path, is_sorted_index=self._blob.index.sorted)
```
Why was `is_sorted_index=self._blob.index.sorted` added?

karitra replied:
If we know in advance that it would be sorted index written on recovery, it is good checkpoint here to mark index files suffixes while creating blob. It was _jonny-k_ suggestion. 

Should it also be in separate commit?

shaitan replied:
> If we know in advance that it would be sorted index written on recovery, it is good checkpoint here to mark index files suffixes while creating blob. It was _jonny-k_ suggestion.

`self._blob.index.sorted` generally doesn't mean that recovered index will be sorted.

> Should it also be in separate commit?

Yes, both this change and adding `is_sorted_index` to `Blob.create()` should be done in a separate commit.

shaitan commented on eblob_kit.py:
```diff
@@ -1482,16 +1509,19 @@ def fix_blob_command(ctx, path, destination, noprompt):
     else:  # run as child.
-        ctx.obj.setdefault(ReportType.EXTENDED, {})[path] = blob_repairer.stat.as_dict
-        ctx.obj.setdefault(ReportType.BASIC, {})[path] = blob_repairer.stat.as_digest_dict
+        ctx.obj.setdefault(ReportType.BASIC, {})[path] = results_digest
```
What is the purpose of these changes?

shaitan commented on debian/changelog:
```diff
@@ -1,3 +1,9 @@
+eblob-kit (0.1.2) unstable; urgency=medium
+
+  * feat: prohibit distination files overwrite by default
```
Here should be all commits

karitra committed:
test: make tests passable with recent updates

shaitan commented on debian/changelog:
```diff
@@ -1,3 +1,12 @@
+eblob-kit (0.1.2) unstable; urgency=medium
```
Please, update also version in `setup.py`

karitra committed:
refact: creation mode can be set for Blob,IndexFile

karitra committed:
feat: prohibit destination file(s) overwrite by default

karitra committed:
test: check for destination writability

  Mocked is_destination_writable check with different results.

  Methods checked:
   - BlobRepairer.recovery_index static method.
   - BlobRepairer.recovery_blob.
   - BlobRepairer.copy_valid_record.

shaitan commented on tests/test_blob_repairer.py:
```diff
@@ -408,6 +416,74 @@ def test_check_index_non_valid(_mocked_exists,
+                                  _mocked_index_file,
+                                  _mocked_open,
+                                  _mocked_is_writable,
+                                  mocker,
```
`mocker` is unused and can be dropped

shaitan commented on tests/test_blob_repairer.py:
```diff
@@ -408,6 +416,74 @@ def test_check_index_non_valid(_mocked_exists,
+                                  _mocked_is_writable,
+                                  mocker,
+                                  callee):
+    """Check for exception if destination not writable.
```
This docstring should be adopted to the test

shaitan commented on tests/test_blob_repairer.py:
```diff
@@ -408,6 +416,74 @@ def test_check_index_non_valid(_mocked_exists,
+    blob_repairer = BlobRepairer('.')
+
+    assert blob_repairer.valid
+    assert not blob_repairer.stat.index_order_error
```
Why do you check this?

karitra replied:
Copy-pasted, should remove it.

shaitan commented on tests/test_blob_repairer.py:
```diff
@@ -408,6 +416,74 @@ def test_check_index_non_valid(_mocked_exists,
+    assert blob_repairer.valid
+    assert not blob_repairer.stat.index_order_error
+
+    if isinstance(callee, types.FunctionType):
```
Is it a correct condition? What if there will be another static method with different arguments?

karitra replied:
Not sure that correct, but one of the first I've found on SO:
https://stackoverflow.com/questions/8727059/python-check-if-method-is-static
and not selected by TC answer, but with most votes.

In first version of tests, it was separate version for each method, but code was bloated with all tests are almost completely identical, so I ended up with current joint version.

shaitan replied:
How about?
```suggestion
    if callee == BlobRepairer.recover_index:
```

karitra replied:
Yep, that do a trick.

shaitan commented on eblob_kit.py:
```diff
@@ -61,6 +61,17 @@ def dump_to_file(file_name, results):
+def is_destination_writable(src_path, dst_path, overwrite=False):
+    """Check if 'dst_path' file writable.
+
+    TODO(karapuz): tests.
```
TODO is satisfied now?

karitra replied:
Yep, should remove it.

shaitan commented on tests/test_blob_repairer.py:
```diff
@@ -408,6 +416,74 @@ def test_check_index_non_valid(_mocked_exists,
+@mock.patch(OPEN_TO_PATCH, new_callable=mock.mock_open)
+@mock.patch('eblob_kit.IndexFile', autospec=True)
+@mock.patch('eblob_kit.DataFile', autospec=True)
+@mock.patch('os.path.exists', return_value=True)
```
Why is it mocked?

karitra replied:
It is used somewhere within `BlobRepairer` on construction. On inner Blob creation.

shaitan replied:
Could you please find this place? Because `os.path.exists` is called in several places and some of them looks like shouldn't be unconditionally True.

karitra replied:
Please refer to `Blob.__init__`:
https://github.com/interiorem/eblob_kit/blob/4a07d68582b52cf5e8c8c3e4331a71d5aa7423e6/eblob_kit.py#L524

karitra replied:
And seems that it doesn't resonate with code being tested.

shaitan commented on tests/test_blob_repairer.py:
```diff
@@ -408,6 +416,67 @@ def test_check_index_non_valid(_mocked_exists,
+@mock.patch('eblob_kit.DataFile', autospec=True)
+@mock.patch('os.path.exists', return_value=True)
+def test_fix_destination_not_writable(_mock_exist,
+                                      mocked_data_file,
```
```suggestion
                                      _mocked_data_file,
```

shaitan commented on tests/test_blob_repairer.py:
```diff
@@ -408,6 +416,67 @@ def test_check_index_non_valid(_mocked_exists,
+@mock.patch('eblob_kit.DataFile', autospec=True)
+@mock.patch('os.path.exists', return_value=True)
+def test_fix_destination_writable(_mock_exist,
+                                  mocked_data_file,
```
```suggestion
                                  _mocked_data_file,
```

shaitan reviewed (dismissed):
LGTM
